constraints/miden-vm/stack.air

```diff
@@ -396,320 +493,396 @@ ev is_unchanged(main: [values[4]]):
+        dup1([stack]) when op_flags[50]
+        dup2([stack]) when op_flags[51]
+        dup3([stack]) when op_flags[52]
+        dup4([stack]) when op_flags[53]
+        dup5([stack]) when op_flags[54]
+        dup6([stack]) when op_flags[55]
+        dup7([stack]) when op_flags[56]
+        dup9([stack]) when op_flags[57]
+        dup11([stack]) when op_flags[58]
+        dup13([stack]) when op_flags[59]
+        dup15([stack]) when op_flags[60]
+        advpop([stack]) when op_flags[61]
+        sdepth([stack, bookkeeping]) when op_flags[62]
+        clk([stack], [clk]) when op_flags[63]
+
+
 
 # No stack shift operations
 
-ev noop(main: [s[16]], aux: [b[7]]):
-    enf f_dup(b) * delta(elem) = 0 for elem in stack
+ev noop(main: [s[16]]):
+    enf delta(elem) = 0 for elem in s
 
-ev eqz(main: [s[16], helper], aux: [b[7]]):
-    enf f_eqz(b) * (s[0]' * s[0]) = 0
-    enf f_eqz(b) * (s[0]' - (1 - s[0] * helper))
-    enf f_eqz(b) * delta(s[i]) = 0 for i in 1..16
+ev eqz(main: [s[16], h0]):
+    enf s[0]' * s[0] = 0
+    enf s[0]' = 1 - s[0] * h0
+    enf delta(s[i]) = 0 for i in 1..16
 
-ev neg(main: [s[16]], aux: [b[7]]):
-    enf f_neg(b) * (s[0]' + s[0]) = 0
-    enf f_neg(b) * delta(s[i]) = 0 for i in 1..16
+ev neg(main: [s[16]]):
+    enf s[0]' + s[0] = 0
+    enf delta(s[i]) = 0 for i in 1..16
 
-ev inv(main: [s[16]], aux: [b[7]]):
-    enf f_inv(b) * (1 - s[0]' * s[0]) = 0
-    enf f_inv(b) * delta(s[i]) = 0 for i in 1..16
+ev inv(main: [s[16]]):
+    enf s[0]' * s[0] = 1
+    enf delta(s[i]) = 0 for i in 1..16
 
-ev incr(main: [s[16]], aux: [b[7]]):
-    enf f_incr(b) * (s[0]' - s[0] - 1) = 0
-    enf f_incr(b) * delta(s[i]) = 0 for i in 1..16
+ev incr(main: [s[16]]):
+    enf s[0]' = s[0] + 1
+    enf delta(s[i]) = 0 for i in 1..16
 
-ev not(main: [s[16]], aux: [b[7]]):
-    enf f_not(b) * binary_check_sub(s[0])
-    enf f_not(b) * (s[0]' - !s[0]) = 0
-    enf f_not(b) * delta(s[i]) = 0 for i in 1..16
+ev not(main: [s[16]]):
+    enf is_binary([s[0]])
+    enf s[0]' = !s[0]
+    enf delta(s[i]) = 0 for i in 1..16
 
-ev fmpadd(main: [s[16]], aux: [b[7], fmp]):
-    enf f_fmpadd(b) * (s[0]' - s[0] - fmp) = 0
-    enf f_fmpadd(b) * delta(s[i]) = 0 for i in 1..16
+ev fmpadd(main: [s[16]], aux: [fmp]):
+    enf s[0]' = s[0] + fmp
+    enf delta(s[i]) = 0 for i in 1..16
 
 # WARNING: not sure that I got the constraint right
 # What are the "helper" columns? Is it just a different name for hasher decoder columns, or they 
 # are unique for i/o ops?
-ev mload(main: [s[16]], aux: [b[7], helper[3], clk, b_chip]):
+ev mload(main: [s[16]], aux: [helper[3], clk, b_chip]):
     let v = alpha[5] * s[0]' + sum([alpha[i + 5] * helper[3 - i]'])
     let op_mem_read = 12
     let u_mem = alpha[0] + alpha[1] * op_mem_read + alpha[3] * s[0] + alpha[4] * clk + v
-    enf f_mload(b) * (b_chip' * u_mem - b_chip) = 0
-    enf f_mload(b) * delta(s[i]) = 0 for i in 1..16
+    enf b_chip' * u_mem = b_chip
+    enf delta(s[i]) = 0 for i in 1..16
```

Comment:
Similar points to the previous comment: `helper` and `clk` columns are a part of the main trace (not auxiliary trace). Also, referencing random values should be done with `$` prefix. I would rewrite this like this:
```
ev op_mload(main: [s[16], helpers[3], clk], aux: [b_chip]):
    let v = $alpha[5] * s[0]' + sum([$alpha[i + 5] * helper[3 - i]']) # this is missing for part of the statement
    let op_mem_read = 12
    let u_mem = $alpha[0] + $alpha[1] * op_mem_read + $alpha[3] * s[0] + $alpha[4] * clk + v
    enf b_chip' * u_mem = b_chip
    enf s[i]' = s[i] for i in 1..16 # this is missing constraint against s[0]
```

Reply:
I'm not sure what constraint we should apply to the `s[0]`. For `MLOAD` we check memory access correctness and no change starting from position 1.